### PR TITLE
1482/feat/complexity reasons

### PR DESCRIPTION
**Pull request by chenFriedman**

### CODE REVIEW

File: client/src/Utils/ControllerHooks/useExposuresSaving.ts [hunk 1]
```diff
@@ -63,7 +90,7 @@ const useExposuresSaving = () => {
             filteredExposures = (filteredExposures as Exposure[]).map(extractExposureData);
         }
         saveExposureAndFlightDataLogger.info('launching the server request', Severity.LOW);
-
+        checkUpdateInvestigationReasonId(filteredExposures)
         return axios.post('/exposure/updateExposures', {
```
Comment (y-golde) [inline client/src/Utils/ControllerHooks/useExposuresSaving.ts:94]: is update exposures dependent on the reason Id?

File: client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts [hunk 2]
```diff
@@ -35,6 +35,7 @@ const usePersonalInfoTab = (): usePersonalInfoTabOutcome => {
     const epidemiologyNumber = useSelector<StoreStateType, number>(state => state.investigation.epidemiologyNumber);
     const investigationStatus = useSelector<StoreStateType, InvestigationStatus>((state) => state.investigation.investigationStatus);
     const investigatedPatientId = useSelector<StoreStateType, number>(state => state.investigation.investigatedPatient.investigatedPatientId);
+    const complexityReasonsId = useSelector<StoreStateType, any>((state) => state.investigation.complexReasonsId);
```
Comment (y-golde) [inline client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts:38]: why any?

File: client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts [hunk 3]
```diff
@@ -156,6 +157,50 @@ const usePersonalInfoTab = (): usePersonalInfoTabOutcome => {
...
+            .catch((err) => {console.log(err)})
+    }
+
+    const removeInvestigationReasonId = (epidemiologyNumber: number, oldComplexityReasonId: number) => {
+        axios.post('/investigationInfo/deleteComplexityReason', {
+            epidemiologyNumberInput: epidemiologyNumber,
+            oldComplexityReasonId: oldComplexityReasonId})
+            .then(() => {})
+            .catch((err) => {console.log(err)})
+    }
```
Comment (y-golde) [inline client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts:174]: add a logger and maybe move those functions to a hook? those seem similar to the functions in useExposuresSaving

File: client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts [hunk 4]
```diff
@@ -156,6 +157,50 @@ const usePersonalInfoTab = (): usePersonalInfoTabOutcome => {
...
+        if (personalInfoData.institutionName == ''  && complexityReasonsId.includes(7)) {
+            removeInvestigationReasonId(epidemiologyNumber, 7)
+        }
+        if (personalInfoData.relevantOccupation !== "מערכת החינוך"  && complexityReasonsId.includes(8)) {
+            removeInvestigationReasonId(epidemiologyNumber, 8)
+        }
+        if (personalInfoData.relevantOccupation !== "מערכת הבריאות"  && complexityReasonsId.includes(9)) {
+            removeInvestigationReasonId(epidemiologyNumber, 9)
+        }
+    }
```
Comment (y-golde) [inline client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts:202]: move it to the hook too

File: client/src/components/App/Content/LandingPage/InvestigationTable/InvestigationIndicatorsColumn/InvestigationIndicatorsColumn.tsx [hunk 5]
```diff
@@ -9,31 +11,34 @@ import SelfInvestigationIcon from './SelfInvestigationIcon/SelfInvestigationIcon
 const complexInvestigationMessage = 'חקירה מורכבת';
 
 const InvestigationIndicatorsColumn = (props: Props) => {
-    const { isComplex, wasInvestigationTransferred, transferReason , isSelfInvestigated ,selfInvestigationStatus, selfInvestigationUpdateTime} = props;
+    const { isComplex, wasInvestigationTransferred, transferReason, isSelfInvestigated, selfInvestigationStatus, selfInvestigationUpdateTime, complexityReasonsId } = props;
+    const allComplexReasons = useSelector<StoreStateType, (number|null)[]>(state => state.complexReasons);
+    const investigationComplexityReasons = !!complexityReasonsId && complexityReasonsId.map((id) => !!id && allComplexReasons[id - 1]).toString()
```
Comment (y-golde) [inline client/src/components/App/Content/LandingPage/InvestigationTable/InvestigationIndicatorsColumn/InvestigationIndicatorsColumn.tsx:16]: why !! and not `Boolean() / ??` ?

File: client/src/components/App/Content/LandingPage/InvestigationTable/InvestigationIndicatorsColumn/InvestigationIndicatorsColumn.tsx [hunk 6]
```diff
@@ -42,10 +47,11 @@ const InvestigationIndicatorsColumn = (props: Props) => {
 interface Props {
     wasInvestigationTransferred: boolean;
     isComplex: boolean;
+    complexityReasonsId: Array<number | null>;
```
Comment (y-golde) [inline client/src/components/App/Content/LandingPage/InvestigationTable/InvestigationIndicatorsColumn/InvestigationIndicatorsColumn.tsx:50]: change to (number|null)[]

File: client/src/components/App/Content/LandingPage/InvestigationTable/InvestigationTableRow/InvestigationTableRow.tsx [hunk 7]
```diff
@@ -58,6 +58,7 @@ const RowTooltip = (props: RowTooltipProps) => {
 };
 
 interface Props {
+    complexityReasonsId: Array<number | null>;
```
Comment (y-golde) [inline client/src/components/App/Content/LandingPage/InvestigationTable/InvestigationTableRow/InvestigationTableRow.tsx:61]: change to (number | null)[]

File: client/src/components/App/Content/LandingPage/InvestigationTable/useInvestigationTable.ts [hunk 8]
```diff
@@ -50,6 +52,7 @@ export const createRowData = (
     epidemiologyNumber: number,
     validationDate: string,
     isComplex: boolean,
+    complexityReasonsId: Array<number | null>,
```
Comment (y-golde) [inline client/src/components/App/Content/LandingPage/InvestigationTable/useInvestigationTable.ts:55]: change to (number | null)[]

File: client/src/components/App/Content/LandingPage/InvestigationTable/useInvestigationTable.ts [hunk 9]
```diff
@@ -546,7 +570,16 @@ const useInvestigationTable = (parameters: useInvestigationTableParameters): use
...
-    const onInvestigationRowClick = (investigationRow: { [T in keyof IndexedInvestigationData]: any }) => {
+    const onInvestigationRowClick = async (investigationRow: { [T in keyof IndexedInvestigationData]: any }) => {
+        const epidemiologyNum :number = investigationRow.epidemiologyNumber
+        await axios.get('/investigationInfo/getComplexityReason/'+ epidemiologyNum)
+            .then((result) => {
+                if (result?.data && result.headers['content-type'].includes('application/json')) {
+                    setComplexReasonsId(result.data)
+                } else { setComplexReasonsId([]) }
+            })
+            .catch((err) => {})
```
Comment (y-golde) [inline client/src/components/App/Content/LandingPage/InvestigationTable/useInvestigationTable.ts:581]: add logger or move to global function/hook

File: client/src/models/InvestigationTableRow.ts [hunk 10]
```diff
@@ -7,6 +7,7 @@ interface InvestigationTableRow {
     epidemiologyNumber: number;
     validationDate: string;
     isComplex: boolean;
+    complexityReasonsId: Array<number | null>;
```
Comment (y-golde) [inline client/src/models/InvestigationTableRow.ts:10]: chage to (number|null)[];

File: server/src/ClientToDBAPI/GroupedInvestigationsRoute/utils.ts [hunk 11]
```diff
@@ -4,6 +4,7 @@ import InvestigationMainStatus from '../../Models/InvestigationStatus/Investigat
 interface GroupedInvestigations {
     allInvestigations: {
         nodes: [{
+            complexityReasonsId: Array<number | null>;
```
Comment (y-golde) [inline server/src/ClientToDBAPI/GroupedInvestigationsRoute/utils.ts:7]: change to (number|null)[];

File: server/src/ClientToDBAPI/InvestigationInfo/mainRoute.ts [hunk 12]
```diff
@@ -301,4 +305,50 @@ investigationInfo.get('/groupedInvestigationsId', handleInvestigationRequest, (r
...
+investigationInfo.get('/getComplexityReason/:epidemiologyNumber', (request: Request, response: Response) => {
+    const queryVariables = { epidemiologyNumber: parseInt(request.params.epidemiologyNumber) }
+    return graphqlRequest(GET_INVESTIGATION_COMPLEXITY_REASON_ID, response.locals, queryVariables)
+    .then((result) => {
+        response.send(result.data.investigationByEpidemiologyNumber.complexityReasonsId);
+    })
+    .catch(error => {
+        response.status(errorStatusCode).send(error);
+    })
+});
```
Comment (y-golde) [inline server/src/ClientToDBAPI/InvestigationInfo/mainRoute.ts:351]: add logs

File: server/src/ClientToDBAPI/LandingPageRoute/utils.ts [hunk 13]
```diff
@@ -6,6 +6,7 @@ import InvestigationMainStatus from '../../Models/InvestigationStatus/Investigat
 interface UserInvestigations {
     orderedInvestigations: {
         nodes: [{
+            complexityReasonsId: Array<number | null>;
```
Comment (y-golde) [inline server/src/ClientToDBAPI/LandingPageRoute/utils.ts:9]: (number|null)[];

File: server/src/ClientToDBAPI/LandingPageRoute/utils.ts [hunk 14]
```diff
@@ -60,6 +61,7 @@ interface UserInvestigations {
 interface GroupIvestigations {
     orderedInvestigations: {
         nodes: [{
+            complexityReasonsId: Array<number | null>;
```
Comment (y-golde) [inline server/src/ClientToDBAPI/LandingPageRoute/utils.ts:64]: (number|null)[];

**Review by y-golde — CHANGES_REQUESTED** [review]

File: client/src/Utils/ControllerHooks/useExposuresSaving.ts [hunk 15]
```diff
@@ -29,6 +29,7 @@ interface DBExposure extends Omit<Exposure, 'exposureAddress'> {
 const useExposuresSaving = () => {
     const epidemiologyNumber = useSelector<StoreStateType, number>(state => state.investigation.epidemiologyNumber);
     const investigatedPatientId = useSelector<StoreStateType, number>(state => state.investigation.investigatedPatient.investigatedPatientId);
+    const complexityReasonsId = useSelector<StoreStateType, any>((state) => state.investigation.complexReasonsId);
```
Comment (Noam6421) [inline client/src/Utils/ControllerHooks/useExposuresSaving.ts:32]: change any to the appropriate type

File: client/src/Utils/ControllerHooks/useExposuresSaving.ts [hunk 16]
```diff
@@ -42,11 +43,37 @@ const useExposuresSaving = () => {
         });
     } 
 
+    const updateInvestigationReasonId = (epidemiologyNumber: number, newComplexityReasonId: number) => {
+        axios.post('/investigationInfo/updateComplexityReason', {
+            epidemiologyNumberInput: epidemiologyNumber,
+            newComplexityReasonId: newComplexityReasonId})
+            .then(() => {})
+            .catch((err) => {})
```
Comment (Noam6421) [inline client/src/Utils/ControllerHooks/useExposuresSaving.ts:51]: add log

Comment (Noam6421) [reply client/src/Utils/ControllerHooks/useExposuresSaving.ts:51]: **logger ' you can look at similar requests and see how to set it up

File: client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts (hunk 2 above)
Comment (Noam6421) [inline client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts:38]: change any

File: client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts [hunk 17]
```diff
@@ -156,6 +157,50 @@ const usePersonalInfoTab = (): usePersonalInfoTabOutcome => {
 
     const clearSubOccupations = () => setSubOccupations([]);
 
+    const updateInvestigationReasonId = (epidemiologyNumber: number, newComplexityReasonId: number) => {
+        axios.post('/investigationInfo/updateComplexityReason', {
+            epidemiologyNumberInput: epidemiologyNumber,
+            newComplexityReasonId: newComplexityReasonId})
+            .then(() => {})
+            .catch((err) => {console.log(err)})
```
Comment (Noam6421) [inline client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts:165]: change console/log to logger

File: client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts [hunk 18]
```diff
@@ -156,6 +157,50 @@ const usePersonalInfoTab = (): usePersonalInfoTabOutcome => {
...
+    }
+
+    const removeInvestigationReasonId = (epidemiologyNumber: number, oldComplexityReasonId: number) => {
+        axios.post('/investigationInfo/deleteComplexityReason', {
+            epidemiologyNumberInput: epidemiologyNumber,
+            oldComplexityReasonId: oldComplexityReasonId})
+            .then(() => {})
+            .catch((err) => {console.log(err)})
```
Comment (Noam6421) [inline client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/usePersonalInfoTab/index.ts:173]: same

File: client/src/components/App/Content/LandingPage/InvestigationTable/InvestigationTable.test.tsx [hunk 19]
```diff
@@ -54,9 +69,24 @@ describe('investigationTable tests', () => {
...
+            newRowData.subOccupation,
+            newRowData.parentOccupation,
+            newRowData.isInInstitute,
+            newRowData.isSelfInvestigated,
+            newRowData.selfInvestigationUpdateTime
+
         );
 
-        expect(newRow).toEqual(newRowData);
+        expect(newRow).not.toEqual(newRowData);
```
Comment (Noam6421) [inline client/src/components/App/Content/LandingPage/InvestigationTable/InvestigationTable.test.tsx:89]: why not?

File: server/src/ClientToDBAPI/InvestigationInfo/mainRoute.ts [hunk 20]
```diff
@@ -301,4 +305,50 @@ investigationInfo.get('/groupedInvestigationsId', handleInvestigationRequest, (r
         })
 });
 
+investigationInfo.get('/complexityReasons', (request: Request, response: Response) => {
```
Comment (Noam6421) [inline server/src/ClientToDBAPI/InvestigationInfo/mainRoute.ts:308]: add logger

**Review by Noam6421 — CHANGES_REQUESTED** [review]

File: client/src/Utils/ControllerHooks/useExposuresSaving.ts (hunk 1 above)
Comment (chenFriedman) [reply client/src/Utils/ControllerHooks/useExposuresSaving.ts:94]: no

File: server/src/ClientToDBAPI/InvestigationInfo/mainRoute.ts (hunk 20 above)
Comment (chenFriedman) [reply server/src/ClientToDBAPI/InvestigationInfo/mainRoute.ts:308]: fixed

**Review by y-golde — APPROVED** [review]

**Review by Noam6421 — APPROVED** [review]

**Pull request merged**